nidmviewerfsl/lib/slicerTools.py

```diff
@@ -1,195 +1,87 @@
-#
-#  / 1 0 0 s*dx \
-# |  0 1 0 s*dy  |
-# |  0 0 1  0    |
-#  \ 0 0 0  1   /
-#
-# where s = 1/v_spm where v_spm is the voxel size of the SPM map.
-#
-# ------------------------------------------------------------------------------
-# Source: https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FLIRT/FAQ
-# Author: Tom Maullin (29/11/2017)
+# Authors: Tom Maullin, Camille Maumet (29/11/2017)
 
 import subprocess
 import os
 import shutil
 import random
 import shlex
+import numpy as np
+import nibabel as nib
+from nibabel.processing import resample_from_to
+from queries.queryTools import runQuery
 
 
-def nifDim(niftiFilename, k):
-    # Retrieve the k dimension of a nifti given it's filename using FSL.
+def nifDim(nifti, k):
+    # Retrieve the k dimension of a nifti using nibabel.
+
+    # Retrieve image header.
+    n = nib.load(nifti)
+    header = n.header
 
     if k == 'x':
-        arg = 'dim1'
+        dimension = header['dim'][1]
     elif k == 'y':
-        arg = 'dim2'
+        dimension = header['dim'][2]
     elif k == 'z':
-        arg = 'dim3'
+        dimension = header['dim'][3]
     elif k == 'pix':
-        arg = 'pixdim1'
+        dimension = header['pixdim'][1]
     else:
         error('Enter a valid dimension... x, y, z or pix')
 
-    # Make the commands
-    getDimString1 = ["fslhd", niftiFilename]
-    getDimString2 = ["cat", "-v"]
-    getDimString3 = ["grep", "^" + arg]
-
-    # Run the command
-    process_1 = subprocess.Popen(getDimString1, shell=False,
-                                 stdout=subprocess.PIPE)
-    process_2 = subprocess.Popen(getDimString2, shell=False,
-                                 stdin=process_1.stdout,
-                                 stdout=subprocess.PIPE)
-    process_3 = subprocess.Popen(getDimString3, shell=False,
-                                 stdin=process_2.stdout,
-                                 stdout=subprocess.PIPE)
-
-    # Close all streams and retreive output.
-    process_1.stdout.close()
-    process_2.stdout.close()
-    output = process_3.communicate()
-
-    dimension = int(float(output[0].decode('utf-8').rstrip(
-        '\r|\n').replace(arg, '').replace(' ', '')))
-
     return(dimension)
 
 
-def createResizeMatrix(niftiFilename1, niftiFilename2, scalefactor, tempDir):
-    # This creates the resize matrix for the resizing of the niftis and saves
-    # it as resizeMatrix.mat
+def resizeTemplateOrExcSet(exc_set, template, tempDir):
+    # This function resizes an SPM excursion set to an SPM template if
+    # necessary.
 
-    xShift = abs(nifDim(niftiFilename1, 'x') - nifDim(niftiFilename2, 'x'))
-    yShift = abs(nifDim(niftiFilename1, 'y') - nifDim(niftiFilename2, 'y'))
+    # Load the images
+    template_img = nib.load(template)
+    excset_img = nib.load(exc_set)
 
-    # Open the matrix file.
-    matrixFile = open(os.path.join(tempDir, 'resizeMatrix.mat'), 'w')
+    # Resample if necessary
+    if nifDim(exc_set, 'x') > nifDim(template, 'x'):
```

Comment:
Can we check the need for reslicing using the complete affine matrix (cf. ttp://nipy.org/nibabel/nifti_images.html#the-nifti-affines)? This will ensure that two images with the same size but different orientation get resliced.

Reply:
I am a little confused by this. Are you suggesting we check if the two affine matrices are identical in order to decide whether or not to resample?

Reply:
Yes, that's right! Do you think there is something wrong with using the affine matrices?